File: java/src/main/java/toolgood/algorithm/mathNet/Precision.java
package toolgood.algorithm.mathNet;

import java.math.BigDecimal;
import java.math.MathContext;

public class Precision {
    public static final BigDecimal DecimalPrecision = new BigDecimal("0.0000000000000000000000000001");

    public static final BigDecimal PositiveDecimalPrecision = new BigDecimal("2").multiply(DecimalPrecision);

    private static final BigDecimal DefaultDecimalAccuracy = DecimalPrecision.multiply(new BigDecimal("10"));

    public static boolean AlmostEqualNormRelative(BigDecimal a, BigDecimal b, BigDecimal diff, BigDecimal maximumError) {
        if (a.abs().compareTo(DecimalPrecision) < 0 || b.abs().compareTo(DecimalPrecision) < 0) {
            return diff.abs().compareTo(maximumError) < 0;
        }

        if ((a.compareTo(BigDecimal.ZERO) == 0 && b.abs().compareTo(maximumError) < 0) 
                || (b.compareTo(BigDecimal.ZERO) == 0 && a.abs().compareTo(maximumError) < 0)) {
            return true;
        }

        return diff.abs().compareTo(maximumError.multiply(a.abs().max(b.abs()))) < 0;
    }

    public static boolean AlmostEqualRelative(BigDecimal a, BigDecimal b) {
        return AlmostEqualNormRelative(a, b, a.subtract(b), DefaultDecimalAccuracy);
    }

    public static boolean AlmostEqual(BigDecimal a, BigDecimal b) {
        return AlmostEqualNorm(a, b, a.subtract(b), DefaultDecimalAccuracy);
    }

    public static boolean AlmostEqualNorm(BigDecimal a, BigDecimal b, BigDecimal diff, BigDecimal maximumAbsoluteError) {
        return diff.abs().compareTo(maximumAbsoluteError) < 0;
    }
}

File: csharp/ToolGood.Algorithm/MathNet/Precision.cs
using System;

namespace ToolGood.Algorithm.MathNet.Numerics
{
    internal static partial class Precision
    {
        public static readonly decimal DecimalPrecision = 0.0000000000000000000000000001M;

        public static readonly decimal PositiveDecimalPrecision = 2M * DecimalPrecision;

        private static readonly decimal DefaultDecimalAccuracy = DecimalPrecision * 10M;

        public static bool AlmostEqualNormRelative(this decimal a, decimal b, decimal diff, decimal maximumError)
        {
            if(Math.Abs(a) < DecimalPrecision || Math.Abs(b) < DecimalPrecision) {
                return Math.Abs(diff) < maximumError;
            }

            if((a == 0 && Math.Abs(b) < maximumError) || (b == 0 && Math.Abs(a) < maximumError)) {
                return true;
            }

            return Math.Abs(diff) < maximumError * Math.Max(Math.Abs(a), Math.Abs(b));
        }

        public static bool AlmostEqualRelative(this decimal a, decimal b)
        {
            return AlmostEqualNormRelative(a, b, a - b, DefaultDecimalAccuracy);
        }

        public static bool AlmostEqual(this decimal a, decimal b)
        {
            return AlmostEqualNorm(a, b, a - b, DefaultDecimalAccuracy);
        }

        public static bool AlmostEqualNorm(this decimal a, decimal b, decimal diff, decimal maximumAbsoluteError)
        {
            return Math.Abs(diff) < maximumAbsoluteError;
        }
    }
}
